### ISSUE TITLE
Sketch Out Data Model for Orgs/Opportunities

### ISSUE THREAD

**Issue by timothyfcook** [saxifrage/learn#16]

**Comment by timothyfcook** [comment]
Here's a draft of what the entire scope might eventually look like, including relationships. This was created reference graphDB type relationships.

Note the distinction and relationship between Opportunity and Opportunity instance:
![datamodel](https://cloud.githubusercontent.com/assets/1823300/6757692/6470aa1c-cf09-11e4-8f61-1bbdda46e62c.jpg)

**Comment by timothyfcook** [comment]
Okay. I created a draft of a JSON file for `opportunity_instance`. It includes an `organizer` object (which replaces `organization` with this more appropriate/generic term since the `organizer` could be an individual, or an informal group).

To create it I referenced the existing [City of Learning API](http://www.col-engine-staging.com/apipie/1.0.html), schema.org and, primarily, the [Eventbrite API](http://developer.eventbrite.com/docs/event-details/).

I'm not sure if/how we want to differentiate between `opportunity` and `opportunity_instance`. In the City of Learning platform currently there isn't any direct data relationship, I think it just uses "opportunity" as a template to create multiple instances. If we want, we could do something more interesting in which `opportunity_instance` is an object within an `opportunity`.

In general, I think these data-types are more thorough and better aligned into nested objects than the City of Learning JSON spec. I also included some data-types that the City of Learning group in Dallas had noted as important and absent. Something in this direction is a nice improvement.

Once we finish our data model I'll create a translation that matches up our data-types with the City of Learning API so that we can talk to it when the time comes.

Thoughts/edits?? @matthannigan any high-level feedback?

Here is the file: https://github.com/saxifrage/learn/blob/master/examples/opportunity_instance_example.json

Code also below:

```

{
  "errors": {},
  "metadata": {},
  "result": {
    "id": 957945853,
    "resource_uri": "https://www.cityasacampus.org/opportunities/957945853/",
    "name": {
        "text": "Awesome Event Name", 
        "html": "Awesome <strong>Styled</strong> Event Name"
    }, 
    "description": "This tells you all about the workshop and how cool it's going to be.",
    "organizer_id": 845309360,
    "organizer": {
        "description": {
            "text": "Carpentry Town wants to teach everyone about Carpentry!", 
            "html": "<P><SPAN STYLE=\"font-size: small;\"><SPAN STYLE=\"color: black; font-family: 'Verdana','sans-serif';\">Carpentry Town wants to teach everyone about Carpentry!<\/SPAN><SPAN STYLE=\"color: black; font-family: 'Verdana','sans-serif'; mso-bidi-font-family: Verdana;\">.<\/SPAN><\/SPAN><\/P>\r\n<P>\u00a0<\/P>"
        },
        "logo": {
            "id": "151295", 
            "url": "http://cdn.evbuc.com/images/151295/28467766139/1/logo.jpg"
        },
        "resource_uri": "https://www.cityasacampus.org/organizers/845309360/", 
        "id": "845309360", 
        "name": "Carpentry Town", 
        "url": "http://www.cityasacampus.org/organizers/carpentry_town",
        "external_url": "http://www.carpentrytown.biz",
        "contact": {
            "contact_name": "Bill Woodsmith",
            "contact_email": "bill@carpentrytown.biz",
            "contact_phone": "4129999999",
        }
    },
    "min_age": 14,
    "max_age": 99,
    "venue_name": "Carpentry Town",
    "venue_id": 1060662062,
    "venue": {
        "address": {
            "address_1": "6587 Butler Street", 
            "address_2": null, 
            "city": "Pittsburgh", 
            "region": "PA", 
            "postal_code": "15206", 
            "country": "US", 
            "latitude": 40.458334999999998, 
            "longitude": -79.911413999999979
        },
        "resource_uri": "https://www.cityasacampus.org/venues/1060662062/", 
        "id": "1060662062", 
        "name": "Carpentry Town", 
        "latitude": "40.458335", 
        "longitude": "-79.91141399999998",
        "neighborhood": "Lawrenceville",
        "neighborhood_id": 12
    },
    "online_opportunity": false,
    "ongoing": false,
    "price": 0,
    "url": "https://www.cityasacampus.org/organizers/carpentry_town/opportunities/basic_carpentry",
    "meeting_format": {
      "format_id": 2,
      "language": ["Spanish", "English"],
      "notes": "Enter through the side door of the building, the front door is broken.",
    },
    "registration_deadline": "2015-05-07T02:00:00Z",
    "registration_url": "http://www.carpentrytown.com/basiccarpentry/march2015",
    "external_url": "http://www.carpentrytown.com/basic_carpentry/",
    "start": {
        "timezone": "America/New_York", 
        "local": "2015-03-06T19:00:00", 
        "utc": "2015-03-07T02:00:00Z"
    }, 
    "created": "2015-03-27T22:27:04Z", 
    "changed": "2015-04-20T19:23:14Z",
    "end": {
        "timezone": "America/New_York", 
        "local": "2015-06-06T22:00:00", 
        "utc": "2015-06-07T05:00:00Z"
    },
    "listing_start": {
        "timezone": "America/New_York", 
        "local": "2015-03-06T19:00:00", 
        "utc": "2015-03-07T02:00:00Z"
    },
    "listing_end": {
        "timezone": "America/New_York", 
        "local": "2015-03-06T19:00:00", 
        "utc": "2015-03-07T02:00:00Z"
    },
    "meeting_times": [
      {
        "meeting_id": 1,
        "meeting_start":{
            "timezone": "America/New_York", 
            "local": "2015-03-06T22:00:00", 
            "utc": "2015-03-07T05:00:00Z"
        },
        "meeting_end":{
            "timezone": "America/New_York", 
            "local": "2015-03-06T22:00:00", 
            "utc": "2015-03-07T05:00:00Z"
        }
      },
      {
        "meeting_id": 2,
        "meeting_start":{
            "timezone": "America/New_York", 
            "local": "2015-04-06T22:00:00", 
            "utc": "2015-04-07T05:00:00Z"
        },
        "meeting_end":{
            "timezone": "America/New_York", 
            "local": "2015-04-06T22:00:00", 
            "utc": "2015-04-07T05:00:00Z"
        }
      },
      {
        "meeting_id": 3,
        "meeting_start":{
            "timezone": "America/New_York", 
            "local": "2014-05-06T22:00:00", 
            "utc": "2014-05-07T05:00:00Z"
        },
        "meeting_end":{
            "timezone": "America/New_York", 
            "local": "2015-06-06T22:00:00", 
            "utc": "2015-06-07T05:00:00Z"
        }
      },
    ],
    "listed": true,
    "category": {
        "resource_uri": "https://www.cityasacampus.org/v1/categories/102/", 
        "id": "102",
        "name": "Fabrication and Building", 
        "short_name": "Fabrication", 
    }, 
    "subcategory": {
        "resource_uri": "https://www.cityasacampus.org/v1/subcategories/2002/", 
        "id": "2002", 
        "name": "Building Trades"
    },
    "badges": [],
  },
  "status": 201
}

```

**Comment by MatthewVita** [comment]
@timothyfcook 

:+1: for the switch from "organization" to "organizer" (very good idea!)

:-1: for putting HTML in the JSON response. Clients should do the styling unless there are _specific cases_ where styling must be enforced from the backend/API. In those cases, I suggest markdown since it's client agnostic. At the moment, our client is a web browser but consider if non-web clients were to emerge (think Android/Java app).

:-1: for including `status` in the JSON response. Statuses should be set at the header level... for example, in Express.js one would do something like:

``` javascript
srv.get('/api/v1/foo/:bar', function(req, res){
  if (req.params.bar) {
    return res.status(200).json({hello: 'world'});
  }

  return res.status(500).json({err: 'you dun goofed, son'});
});
```

TL;DR: `status` is redundant.

...2 questions:
- What is the point of the `metadata` field?
- Why do we need an `errors` field in cases where there are no errors? Shouldn't it only be present if there's an error?

(Great job so far!)

**Comment by timothyfcook** [comment]
Cool, great feedback. The `status`, `metadata`, `errors`, fields are holdovers from the City of Learning JSON responses. I included them just in case anyone thought they were useful. The HTML versions of the names (e.g. `name.html`) is, interestingly, something that Eventbrite provides in their JSON responses, I left it there from their reference in case anyone thought it was useful.

Maybe the html/styling has some use in terms of the API...? I'm sure they have some reason...

**Comment by timothyfcook** [comment]
Okay, loaded in the Eventbrite and City of Learning JSON responses as examples, here: https://github.com/saxifrage/learn/tree/master/examples

**Comment by MatthewVita** [comment]
@timothyfcook @justinxreese what do you guys think of this? This is a branch on my fork that changes `organizations` to `organizers`:

https://github.com/MatthewVita/learn/commit/f3f96c7b845cdf51c658d7e1f9b5ec75896c1671

**Comment by justinxreese** [comment]
Instead of editing the migrations, you should write new ones. It'll look
like this

Use the command `rails g migration`

Edit the file it creates to something like this

``` ruby
class RenameOldTableToNewTable< ActiveRecord::Migration
def change
rename_table :old_table_name, :new_table_name
end
end
```

Then do another one to rename the associations in the related tables.

Excuse weirdness in formatting. Sending from my phone.

**Comment by timothyfcook** [comment]
Re-worked that initial draft to begin referencing json-ld: https://github.com/saxifrage/learn/blob/master/examples/opportunity_instance_linked_data.json

**Comment by timothyfcook** [comment]
Moved to #50

**Comment by timothyfcook** [comment]
@matthannigan we're going to go ahead and start implementing this in #50 ...feel free to leave make any suggestions in #45 or in the [dev.chat](https://gitter.im/saxifrage/learn)
